src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,131 @@
+mod event;
+mod notification;
+
+use event::LogEvent;
+
+use crate::api::{
+    tap::{TapController, TapNotification, TapResult, TapSink},
+    ControlSender,
+};
+use async_graphql::{validators::IntRange, Context, Subscription, Union};
+use futures::StreamExt;
+use itertools::Itertools;
+use tokio::{select, stream::Stream, sync::mpsc, time};
+
+#[derive(Union)]
+/// Log event result which can be a log event or notification
+pub enum LogEventResult {
+    /// Log event
+    LogEvent(event::LogEvent),
+    /// Notification
+    Notification(notification::LogEventNotification),
+}
+
+/// Convert an `api::TapResult` to the equivalent GraphQL type.
+impl From<TapResult> for LogEventResult {
+    fn from(t: TapResult) -> Self {
+        use notification::{LogEventNotification, LogEventNotificationType};
+
+        match t {
+            TapResult::LogEvent(name, ev) => Self::LogEvent(LogEvent::new(&name, ev)),
+            TapResult::Notification(name, n) => match n {
+                TapNotification::Matched => Self::Notification(LogEventNotification::new(
+                    &name,
+                    LogEventNotificationType::Matched,
+                )),
+                TapNotification::NotMatched => Self::Notification(LogEventNotification::new(
+                    &name,
+                    LogEventNotificationType::NotMatched,
+                )),
+            },
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct LogEventsSubscription;
+
+#[Subscription]
+impl LogEventsSubscription {
+    /// A stream of log events emitted from matched component(s)
+    pub async fn output_log_events<'a>(
+        &'a self,
+        ctx: &'a Context<'a>,
+        component_names: Vec<String>,
+        #[graphql(default = 500)] interval: i32,
+        #[graphql(default = 100, validator(IntRange(min = "1", max = "10_000")))] limit: i32,
+    ) -> impl Stream<Item = Vec<LogEventResult>> + 'a {
+        let control_tx = ctx.data_unchecked::<ControlSender>().clone();
+        create_log_events_stream(
+            control_tx,
+            &component_names,
+            // GraphQL only supports 32 bit ints out-the-box due to JSON limitations; we're
+            // casting here to separate concerns and avoid 64 bit scalar deserialization.
```

Comment:
I'm wondering if we should be using `u64::try_from()`, and friends, and bubbling up errors. Right now this will silently convert, for example, `-1` into `18446744073709551615`

Reply:
This shouldn't happen in practice, since the following annotation restricts values to the 1 - 10,000 range:

```rust
#[graphql(default = 100, validator(IntRange(min = "1", max = "10_000")))]
```

If that rule were to change, however, it _could_ wrap. It looks like async-graphql has a `u32` implementation, which would always fit within u64, and I think offers a more realistic upper bound. Will switch.